### PR TITLE
Safe routers epoch update

### PR DESCRIPTION
**Pull request by rogermichoud**
Upon invalidateLayout, we should only set the client routers epoch to the
new layout epoch if they are greater than the current one they have. There
is no reason to get the routers to a smaller epoch. This prevents a downgrading
of epoch between seal and installation of a new Layout.

### CODE REVIEW

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ff2a6fd**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #822 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/m4oRFC7ZUN279aHVt3L94nDrGK9nUOSU)

**Comment by rogermichoud** [comment]
Some test will be stuck in infinite loop. These tests are dependent on #834 to pass. Tested with cherry-picked #834 and all the tests pass.

This is ready for review.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **7b1b738**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #822 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/N4nO0hzyrBZZ9aEao5xbFuWfjIOi1j0F)

**Comment by rogermichoud** [comment]
@dahliamalkhi  I added the check in setEpoch for both NettyClientRouter and TestClientRouter. I also added test for both of them.

**Comment by corfudb-performance** [comment]
Results automatically generated by _CorfuDB Benchmark Framework_ to assess the performance of this pull request for commit **ce97079**.

*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Scan workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx+Iterator workload, 10 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 1 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 5 threads, Disk mode
*** 0.0% transaction FAILURE rate for NonConflictingTx workload, 10 threads, Disk mode
 
 An interactive dashboard with Pull Request Performance Metrics for ALL cluster types and numbers of threads in run, is available at:
 [Pull Request #822 Graphs ](https://snapshot.raintank.io/dashboard/snapshot/uGi06gX4aK2XuU4vsxMn550ycnUxG09u)

File: runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java [hunk 1]
```diff
@@ -82,8 +82,22 @@
...
     public long epoch;
+
+    /**
+     * We should never set epoch backwards
+     *
+     * @param epoch
+     */
+    public void setEpoch(long epoch){
+        if (epoch < this.epoch) {
+            log.warn("setEpoch: attempt to set the router {}:{} to epoch {} smaller than current epoch {}",
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: Can we rephrase this as "setEpoch: Rejecting attempt"

Comment (rogermichoud) [inline runtime/src/main/java/org/corfudb/runtime/clients/NettyClientRouter.java]: Done

**Comment by slfritchie** [comment]
Roger's comment about the failing test case (infinite loop) at https://github.com/CorfuDB/CorfuDB/pull/822#issuecomment-315500044 is correct: merging #834 (at commit 2be5ddeb6d1265612930ec8447f4600079866ac9) into this branch will fix the `ManagementViewTest` tests.
